contrib/src/main/java/gov/nasa/jpl/aerie/spice/SpiceLoader.java

```diff
@@ -12,24 +12,23 @@ public final class SpiceLoader {
     private SpiceLoader() {}
 
     public static void loadSpice() {
-        final String library = "JNISpice";
+        try (final InputStream in = SpiceLoader.class.getResourceAsStream(getResourcePathByOS())) {
+            if (in == null) throw new RuntimeException("JNISpice native library not found");
 
-        // Attempt to load our copy of JNISpice via `System.load`.
-        try (final InputStream in = SpiceLoader.class.getResource(getResourcePathByOS()).openStream()) {
             // Copy the JNISpice library to a temporary location on-disk, then load it into the JVM.
-            final Path file = Files.createTempFile(library + "-", ".tmp");
-            deleteFileOnExit(file);
+            // This needs to be a distinct file from any other load, or else the OS will happily give
```

Comment:
Vague concern that I don't have a concrete suggestion for:

It doesn't seem like we're guaranteeing that this name is unique (and in general, we probably _can't_ guarantee it). Specifically I'm considering a naive implementation of `createTempFile` such as:
```python
def createTempFile(name):
    i = 0
    while exists(name + i):
        i += 1
    return makeFile(name + i)
```

I suppose the answer is "most createTempFile implemenations don't do that, so stop worrying, Matt".

Reply:
Ah, but I delete the file afterwards, so it *must* be a unique file (inode) every time, regardless of the name 😇

Reply:
To be more concrete: it's not the *name* that must be unique, it's the *file* (i.e. the inode representing that file in the filesystem). Naming files different things at different paths is simply one way to make sure you're always creating a new file and not accidentally overwriting the contents of another. But since I delete the file immediately after it's loaded, the inode remains but is totally inaccessible from the filesystem. So we're always creating new files, every time.

Reply:
Also, Javadocs say this:

> Creates a new empty file in the specified directory, using the given prefix and suffix strings to generate its name. If this method returns successfully then it is guaranteed that:
> 1. The file denoted by the returned abstract pathname did not exist before this method was invoked, and
> 2. Neither this method nor any of its variants will return the same abstract pathname again in the current invocation of the virtual machine.

Which is quite cool.

Reply:
That is quite cool! In that case, my concern is assuaged.

I hadn't considered that it might be the inode that needed to be unique - definitely could use some brushing-up on filesystems.

What led me to worry about the uniqueness of the filename was this snippet of code:

([source](https://github.com/openjdk/jdk18/blob/0f2113cee79b9645105b4753c7d7eacb83b872c2/src/java.base/share/classes/jdk/internal/loader/NativeLibraries.java#L196-L206))
```java
// find if this library has already been loaded and registered in this NativeLibraries
NativeLibrary cached = libraries.get(name);
if (cached != null) {
    return cached;
}


// cannot be loaded by other class loaders
if (loadedLibraryNames.contains(name)) {
    throw new UnsatisfiedLinkError("Native Library " + name +
            " already loaded in another classloader");
}
```

Now, I haven't done any digging into where `loadedLibraryNames` comes from - however, I am satisfied by guarantee number 2 above - "Neither this method nor any of its variants will return the same abstract pathname again in the current invocation of the virtual machine."

Reply:
Oh fascinating. When I did this with my custom JNI prototype, I was able to get away with not renaming the library file at all -- just putting it in a different directory was sufficient. 

(Also, look at you, linking to the OpenJDK sources on GitHub 👀)

Okay, so I think I need to update the factors listed in the summary. *Both* the inode *and* the name must be unique -- the inode so the OS doesn't cheat, and the filename so Java doesn't scream. (You can have the same inode with a different name using a hard link; this causes the OS to cheat and say "no, I know I already opened this, you can have a new handle to it though.")

I really thought inode uniqueness would be sufficient. Huh.

Reply:
Yeah, the caller (in our flow from `System.load` anyway) of the function you linked computes `name` as `file.getCanonicalPath()`, so the name is sensitive to the entire path, not just the basename. That explains why both my directory name scheme and the `SpiceLoader`'s random temp filename scheme get past this requirement.